### PR TITLE
feat: webui v2 routes with RebornServicesApi

### LINKED ISSUE #3611: [Reborn WebUI Beta] Implement minimal native WebChat v2 routes
Parent tracker: #3607
Module: M1 — WebUI Product Surface
Priority: P0

## Goal

Implement the minimal default-off WebChat v2 native Reborn route set for beta. This is a native WebUI surface, not a ProductAdapter/WASM surface.

## Route scope

- Create thread.
- Send message.
- Get timeline/snapshot.
- Stream timeline/progress.
- Cancel run if supported by the beta loop.
- Resolve gate only if approval/auth gates are included in beta.

## Acceptance criteria

- Routes preserve WebUI auth/security invariants: bearer/DB/OIDC as applicable, query-token exceptions, CSRF/origin, CORS, body limits, rate limits, connection limits, static security headers.
- Mutations are HTTP. Streaming is read-only, preferably SSE first.
- Route handlers consume only WebUI-facing `RebornServices`/facades. No raw dispatcher, HostRuntime, run-state, DB store, or runtime-lane access.
- Feature/default-off flag exists.
- Caller-level handler/API tests cover submit and timeline/stream behavior.

## Related

Supersedes the ProductAdapter framing in #3282 and aligns with #3580/#3578.

### PR DESCRIPTION
**Pull request by italic-jinxin**
## Summary

Adds `ironclaw_webui_v2` — the minimal native WebChat v2 HTTP route
surface on top of `RebornServicesApi`. Six axum handlers (`create_thread`,
`send_message`, `get_timeline`, `stream_events` SSE, `cancel_run`,
`resolve_gate`) each ship with an `IngressRouteDescriptor` from #3683
that declares listener class, bearer auth, scope source, body/rate
limit, CORS, streaming mode, audit class, and allowed effect path —
host composition enforces the policy before any handler runs.
Compiled in only behind the `webui-v2-beta` Cargo feature; default-off.
SSE supports `Last-Event-ID` for browser `EventSource` auto-reconnect.
Architecture boundary test forbids direct deps on dispatcher, host
runtime, run state, runtime lanes, storage, or the v1 root. Also
promotes `RebornServices::create_thread` to the `RebornServicesApi`
trait so handlers run against a single `Arc<dyn ...>`. Covered by 5
descriptor contract tests + 9 caller-level handler tests against a
stub facade.

## Dependency

Closes the route surface required by #3611 acceptance criterion #2
("mutations are HTTP, streaming is SSE first") and #3 ("handlers
consume only WebUI-facing facades"). Host composition wiring that
mounts the router under the WebUI gateway listener and runs the
bearer-token middleware lands in a follow-up PR; this PR is reviewable
on its own.

## Change Type

- [ ] Bug fix
- [x] New feature
- [x] Refactor
- [x] Documentation
- [ ] CI/Infrastructure
- [ ] Security
- [ ] Dependencies

## Linked Issue

Closes #3611 

## Validation

- [ ] `cargo fmt --all -- --check`
- [ ] `cargo clippy --all --benches --tests --examples --all-features -- -D warnings`
- [ ] `cargo build`
- [ ] Relevant tests pass: 
- [ ] `cargo test --features integration` if database-backed or integration behavior changed
- [ ] Manual testing: 
- [ ] If a coding agent was used and supports it, `review-pr` or `pr-shepherd --fix` was run before requesting review

## Review Follow-Through

---

**Review track**:

### CODE REVIEW

File: crates/ironclaw_webui_v2/tests/webui_v2_handlers_contract.rs [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+    /// Triggered the first time `stream_events` is invoked. Lets the SSE
+    /// test wait on the actual facade call rather than guessing at a
+    /// timeout — axum's SSE body is lazy, so the handler does not run
+    /// until the client polls the body.
+    fn stream_events_signal(&self) -> Arc<Notify> {
+        self.stream_events_notify.clone()
+    }
+}
+
+#[async_trait]
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/tests/webui_v2_handlers_contract.rs:103]: **High Severity**

`StubServices` no longer implements `RebornServicesApi`, so this crate's own test target does not compile.

`RebornServicesApi` now requires `get_run_state`, but the handler contract test stub omits it. Running `cargo test -p ironclaw_webui_v2 --features webui-v2-beta` fails with:

```text
error[E0046]: not all trait items implemented, missing: `get_run_state`
  --> crates/ironclaw_webui_v2/tests/webui_v2_handlers_contract.rs:76:1
```

Fix by adding a stub `get_run_state` implementation here, or split the handler-facing facade so route tests only need the methods the router actually calls.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+fn build_sse_stream(
+    services: std::sync::Arc<dyn RebornServicesApi>,
+    caller: WebUiAuthenticatedCaller,
+    thread_id: String,
+    initial_cursor: Option<String>,
+) -> impl Stream<Item = Result<Event, Infallible>> {
+    async_stream::stream! {
+        let mut after_cursor = initial_cursor.and_then(parse_cursor_token);
+        loop {
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs:206]: **Medium Severity**

Each open SSE connection polls `services.stream_events` forever once per second.

The route descriptor rate-limits new requests, but it does not cap active streams. An authenticated caller can open many long-lived EventSource connections and multiply backend projection drains at `connections × 1/sec` until clients disconnect. This creates an avoidable per-caller resource-exhaustion path once the router is mounted.

Please add an active per-caller stream cap, max stream lifetime / idle timeout, or require host composition to enforce SSE concurrency before this route is exposed.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+                // Force a clean close so the browser can reconnect with
+                // Last-Event-ID; this caps single-stream lifetime
+                // regardless of client behavior and recycles the slot.
+                return;
+            }
+            let request = RebornStreamEventsRequest {
+                thread_id: thread_id.clone(),
+                after_cursor: after_cursor.clone(),
+            };
+            match services.stream_events(caller.clone(), request).await {
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs]: High severity

This new route exposes `RebornServicesApi::stream_events`, but that facade path currently parses `thread_id` and drains the projection stream with `caller.turn_scope(thread_id)` without the ownership probe used by `cancel_run`, `resolve_gate`, and `get_run_state`. Since `TurnScope` has no `owner_user_id`, an authenticated caller sharing tenant/agent/project can request another user's thread id and reach projection reads if the stream trusts the facade boundary. Please make `stream_events` call `assert_thread_owned_by(...)` before `event_stream.drain(...)`, then add a regression where Bob requests Alice's stream and the projection stream is not called.

File: crates/ironclaw_product_workflow/src/reborn_services.rs [hunk 4]
```diff
@@ -109,6 +115,43 @@ impl RebornServices {
...
+        else {
+            return Err(RebornServicesError::internal_invariant());
+        };
+        let thread_id =
+            requested_thread_id.unwrap_or_else(|| generated_thread_id(&caller, &client_action_id));
+        let scope = caller.turn_scope(thread_id.clone());
+        let thread_scope = thread_scope_from_turn_scope(&scope, Some(caller.user_id.clone()))?;
+        let thread = self
+            .thread_service
+            .ensure_thread(EnsureThreadRequest {
```
Comment (serrrfirat) [inline crates/ironclaw_product_workflow/src/reborn_services.rs:149]: Medium severity

Explicit `requested_thread_id` makes create-thread distinguish another user's existing thread from a free id. If `ensure_thread` sees the same tenant/agent/project thread owned by another user, `ThreadScopeMismatch` flows through `map_thread_error` as `409 Conflict`; a non-existent id creates successfully. That gives same-scope callers an existence oracle for guessed thread ids, unlike the ownership probe path below that deliberately collapses mismatches. Please either avoid arbitrary browser-supplied thread ids here or map explicit-id ownership collisions to the same redacted not-authorized/not-found outcome, with a caller-level regression.

File: crates/ironclaw_webui_v2/src/handlers.rs (hunk 3 above)
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs]: Medium severity

`SSE_MAX_LIFETIME` is only checked before awaiting this drain. If `services.stream_events(...)` or the underlying projection drain stalls, the loop never reaches the lifetime check again and the `SseSlot` remains held indefinitely, defeating the advertised max stream lifetime. Please wrap each drain in a timeout based on the remaining lifetime, or `tokio::select!` between the drain future and a deadline so the stream closes and drops the slot on timeout.

File: crates/ironclaw_webui_v2/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    "dep:tokio",
+]
+
+[dependencies]
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+tracing = "0.1"
+
+ironclaw_host_api = { path = "../ironclaw_host_api", version = "0.1.0" }
+ironclaw_product_adapters = { path = "../ironclaw_product_adapters", version = "0.1.0" }
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/Cargo.toml]: Medium severity

This normal dependency lets the route crate reach product-adapter types directly (`ProjectionCursor` in `handlers.rs`), which conflicts with the local rule that handlers consume only `RebornServicesApi` plus the ingress/error vocabulary. It also means the architecture boundary test cannot enforce the documented single-facade boundary. Please move cursor token parsing/formatting behind `ironclaw_product_workflow` as an opaque WebUI-facing helper/type, remove this dependency, and add `ironclaw_product_adapters` to the forbidden list if that boundary is intended.

File: crates/ironclaw_webui_v2/tests/webui_v2_descriptors_contract.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            route.policy().scope_source(),
+            IngressScopeSource::AuthenticatedCaller,
+            "route {} must source scope from the authenticated caller",
+            route.route_id()
+        );
+    }
+}
+
+#[test]
+fn every_route_carries_a_per_caller_rate_limit() {
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/tests/webui_v2_descriptors_contract.rs]: Medium severity

The file-level contract says these tests lock method + pattern and the host-owned ingress policy, but the assertions only cover count/ids, auth, rate-limit scope, streaming mode, and method. A descriptor path could drift from the router mount, or body limits, rate max/window, CORS, audit class, listener class, websocket origin policy, or allowed effect path could change while this still passes. Please make this table-driven by route id and assert the full descriptor policy surface that host composition will rely on.

**Review by serrrfirat — CHANGES_REQUESTED** [review]
Multi-agent review result: requesting changes.

Primary findings:
- High: the newly exposed SSE route reaches `RebornServicesApi::stream_events`, whose facade path drains projection events without the ownership probe used by cancel/resolve/run-state. This can expose another user's thread stream within the same tenant/agent/project scope if the projection stream trusts the facade boundary.
- Medium: explicit `requested_thread_id` creation can leak cross-user thread-id existence through `ThreadScopeMismatch` -> `409`.
- Medium: SSE max lifetime does not bound a hung `stream_events` drain await, so a stuck backend can hold the per-caller slot indefinitely.
- Medium: `ironclaw_webui_v2` directly depends on `ironclaw_product_adapters` for cursor handling, weakening the documented facade-only route boundary.
- Medium: descriptor tests do not actually pin route patterns and most ingress policy values they claim to lock.

I could not run local cargo validation in this thread; this review is based on the PR diff and public branch sources.

**Review by serrrfirat — CHANGES_REQUESTED** [review]
Reviewed current head `30f476e0`. The follow-up commits resolved most of the earlier review: the handler crate compiles, `stream_events` now ownership-probes before projection drain, explicit thread-id collisions redact to `404`, descriptor tests lock the full policy surface, and the direct `ironclaw_product_adapters` dependency is gone. Two things still need addressing before I’d call this ready.

1. **SSE max lifetime still does not bound a stuck facade drain**

`crates/ironclaw_webui_v2/src/handlers.rs` checks `SSE_MAX_LIFETIME` before awaiting `services.stream_events(...)`, but the awaited drain itself is unbounded. If the projection stream stalls, the generator never reaches the lifetime check again and the `SseSlot` remains held despite the documented recovery guarantee in `crates/ironclaw_webui_v2/src/sse_capacity.rs`.

Please wrap each drain in a timeout or `tokio::select!` against the remaining lifetime. A good regression would use a stub facade whose `stream_events` never resolves, drive the SSE body, advance Tokio time past `SSE_MAX_LIFETIME`, and assert the slot is released / a new stream can open.

2. **`Closes #3611` overstates the landed production surface**

The PR body says host composition wiring, bearer middleware, and route mounting land later, but #3611 acceptance criteria include browser auth/security parity. Existing WebUI docs explicitly require query-token auth for browser SSE because `EventSource` cannot set custom headers (`src/channels/web/CLAUDE.md`), and the current gateway allowlist only covers `/api/chat/events`, `/api/logs/events`, and `/api/chat/ws` (`src/channels/web/platform/auth.rs`). This PR’s descriptor hard-locks the new SSE route to bearer auth, but does not make `/api/webchat/v2/threads/{thread_id}/events` browser-auth reachable.

Please change the PR issue link to `Advances #3611` or land the host auth/mounting/query-token path in this PR. A good follow-up test would drive the mounted route through the actual gateway auth middleware with `?token=` and assert the SSE handler receives `WebUiAuthenticatedCaller`.

Verification I ran locally:
- `cargo test -p ironclaw_webui_v2 --features webui-v2-beta`
- `cargo test -p ironclaw_product_workflow --test reborn_services_contract --features test-support`
- `cargo test -p ironclaw_architecture reborn_crate_dependency_boundaries_hold`
- `cargo check -p ironclaw_webui_v2 --no-default-features --lib`

CI checks are green, but GitHub currently reports the PR as merge-conflicting.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+            // stuck projection drain cannot pin the slot past the budget.
+            let remaining = SSE_MAX_LIFETIME.saturating_sub(started_at.elapsed());
+            if remaining.is_zero() {
+                return;
+            }
+            let request = RebornStreamEventsRequest {
+                thread_id: thread_id.clone(),
+                after_cursor: after_cursor.clone(),
+            };
+            match tokio::time::timeout(
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs:220]: **High: SSE loop polls a heavyweight unbounded facade once per second**

Every active SSE connection calls `services.stream_events` once per second. That facade currently ownership-probes through `list_thread_history`, which loads the thread, all messages, and summary artifacts, then returns all pending projection envelopes as a `Vec`. With the default cap of 3 streams per caller, steady-state idle cost becomes active_streams * 1/sec * transcript_size.

Fix: use a WebUI streaming facade that performs a metadata-only ownership check and drains/subscribes with a bounded page size or blocking cursor instead of full-history polling.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+    Ok(Json(response))
+}
+
+/// `GET /api/webchat/v2/threads/{thread_id}/timeline`
+pub async fn get_timeline(
+    State(state): State<WebUiV2State>,
+    Extension(caller): Extension<WebUiAuthenticatedCaller>,
+    Path(thread_id): Path<String>,
+) -> Result<Json<RebornTimelineResponse>, WebUiV2HttpError> {
+    let request = RebornTimelineRequest { thread_id };
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs]: **Medium: Timeline route exposes full thread history without pagination**

The new `GET /timeline` handler forwards directly to `get_timeline`, whose response contains all messages and summary artifacts from `list_thread_history`. A large thread can force a large allocation and JSON serialization on one request; the route rate limit does not bound per-request response size.

Fix: add cursor/limit fields to the timeline request and cap the number of messages/artifacts returned per response.

File: crates/ironclaw_webui_v2/src/sse_capacity.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        self: &Arc<Self>,
+        tenant_id: &TenantId,
+        user_id: &UserId,
+    ) -> Option<SseSlot> {
+        let key = CallerKey {
+            tenant_id: tenant_id.clone(),
+            user_id: user_id.clone(),
+        };
+        let mut state = self.state.lock().expect("SseCapacity state lock poisoned"); // safety: only this module locks; no nested locks; release is infallible
+        let entry = state.entry(key.clone()).or_insert(0);
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/sse_capacity.rs:75]: **Low: Zero SSE capacity leaks caller entries on rejected opens**

`try_acquire` inserts a `HashMap` entry before checking the limit. If host composition uses `WebUiV2State::with_sse_concurrency_limit(..., 0)` to disable SSE, every rejected open stores a zero-count caller key and no `SseSlot` is returned to release it.

Fix: reject `max_per_caller == 0` before calling `entry()`, or model the configured cap as `NonZeroUsize` with an explicit disabled mode.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+                    // recovers even under a stuck projection stream — without
+                    // this bound, an unbounded `.await` on a non-resolving
+                    // facade would pin the slot indefinitely.
+                    tracing::debug!(
+                        target = "ironclaw_webui_v2::sse",
+                        "stream_events drain pending past SSE_MAX_LIFETIME; closing stream"
+                    );
+                    return;
+                }
+                Ok(Ok(response)) => {
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs:239]: **Medium: SSE projection payload shape is untested**

The handler tests open the SSE route and verify that the facade is drained, but none have the facade return projection envelopes and assert the emitted SSE event name, JSON data, cursor `id`, or cursor advancement across polls. That leaves the core browser-facing stream contract unpinned.

Fix: add `stream_events_emits_projection_events_with_cursor_ids` covering single/multiple facade envelopes, SSE event name/data/id, and next-poll `after_cursor` advancement.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+                    // Bound the poll sleep too so we never oversleep past the
+                    // lifetime budget; the top-of-loop check then fires.
+                    let sleep_for = SSE_POLL_INTERVAL
+                        .min(SSE_MAX_LIFETIME.saturating_sub(started_at.elapsed()));
+                    if sleep_for.is_zero() {
+                        return;
+                    }
+                    tokio::time::sleep(sleep_for).await;
+                }
+                Ok(Err(error)) => {
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs:275]: **Medium: SSE facade error event path is untested**

When `RebornServicesApi::stream_events` returns `Err`, this handler emits a redacted SSE `error` event and closes the stream instead of returning an HTTP error. Existing handler tests cover an HTTP `create_thread` error, but not this distinct streaming error path.

Fix: add `stream_events_facade_error_emits_redacted_error_event_and_closes` covering facade `Err` mapped to an SSE error event with `retryable` only and stream termination.

File: crates/ironclaw_webui_v2/src/handlers.rs [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                        match serde_json::to_string(&envelope) {
+                            Ok(payload) => {
+                                let mut event = Event::default().event("projection").data(payload);
+                                if let Some(id) = id {
+                                    event = event.id(id);
+                                }
+                                yield Ok(event);
+                            }
+                            Err(error) => {
+                                tracing::warn!(
```
Comment (serrrfirat) [inline crates/ironclaw_webui_v2/src/handlers.rs]: **Medium: SSE serialization warning violates logging-level rule**

This internal diagnostic uses `tracing::warn!` for a `ProductOutboundEnvelope` serialization failure. The repo rule in `CLAUDE.md` says `info!` and `warn!` output appears in the REPL/TUI and should be avoided for internal diagnostics; this path is not user-facing status.

Fix: change this diagnostic to `tracing::debug!`, or route a redacted user-visible failure through the response contract instead of warning.

File: crates/ironclaw_architecture/tests/reborn_dependency_boundaries.rs [hunk 13]
```diff
@@ -1088,6 +1088,59 @@ fn boundary_rules() -> Vec<BoundaryRule> {
                 "ironclaw_gateway",
             ],
         },
+        BoundaryRule {
+            // WebChat v2 route surface must only reach into Reborn through
+            // the host-facing facade and the ingress vocabulary; anything
+            // that lets a handler touch the dispatcher, runtime lane, run
+            // state, or a storage backend directly would defeat the
+            // single-facade discipline that this crate exists to enforce.
+            crate_name: "ironclaw_webui_v2",
```
Comment (serrrfirat) [inline crates/ironclaw_architecture/tests/reborn_dependency_boundaries.rs:1104]: **Low: WebUI v2 crate is missing from ingress lifecycle guard**

The diff registers `ironclaw_webui_v2` as a WebChat v2 route surface, but `reborn_product_api_crates_do_not_bind_http_ingress` still scans a fixed `reborn_product_api_src_roots` list that does not include `crates/ironclaw_webui_v2/src`. The host-api contract says Product/API crates must not bind sockets or call `axum::serve`; without adding this crate to that scan, the contract fails open for the new route crate.

Fix: add `crates/ironclaw_webui_v2/src` to `reborn_product_api_src_roots`.

**Review by serrrfirat — CHANGES_REQUESTED** [review]
Forced review of current head `ce95e730b02cd4270c262ef2dfa4a22790ae55f1`.

Reviewed with 5 lenses: security, bugs, performance/concurrency, tests, conventions.

Summary:
- Security: 0 findings
- Bugs: 0 findings
- Performance/concurrency: 3 findings
- Tests: 2 findings
- Conventions: 2 findings

Event: REQUEST_CHANGES because at least one High-severity finding has confidence >= 75.

Highest-risk issue: the SSE handler polls `RebornServicesApi::stream_events` once per second per open connection, and that facade currently performs a full history-backed ownership probe before draining events. This makes idle EventSource connections scale with transcript size and active stream count.

**Comment by think-in-universe** [comment]
# PR Review Summary

PR: https://github.com/nearai/ironclaw/pull/3747  
Branch: `feat/webui-v2-routes-3611`

## Findings

- No blocking issues found in the reviewed scope.
- No medium/high-risk regressions identified in auth/ownership checks, SSE lifecycle handling, cursor flow, or facade boundary usage.

## What Was Reviewed

- WebUI v2 route surface and policies in:
  - `crates/ironclaw_webui_v2/src/handlers.rs`
  - `crates/ironclaw_webui_v2/src/descriptors.rs`
  - `crates/ironclaw_webui_v2/src/sse_capacity.rs`
  - `crates/ironclaw_webui_v2/src/router.rs`
- Reborn service facade behavior in:
  - `crates/ironclaw_product_workflow/src/reborn_services.rs`
- Thread-service contract additions and backend implementations in:
  - `crates/ironclaw_threads/src/service.rs`
  - `crates/ironclaw_threads/src/in_memory.rs`
  - `crates/ironclaw_threads/src/filesystem_service.rs`

## Validation Performed

- `cargo test -p ironclaw_webui_v2` (pass)
- `cargo test -p ironclaw_product_workflow --test reborn_services_contract` (pass, 38 tests)

## Residual Risk / Follow-ups

- Full workspace/e2e gateway composition was not run in this pass; integration wiring (feature flags, middleware ordering, host mount composition) remains residual risk.

## Recommendation

- Approve from code-review perspective, contingent on CI and integration checks being green.

**Review by think-in-universe — APPROVED** [review]
